### ISSUE TITLE
[hdpowerview] All things are offline with PowerView Hub v1

### ISSUE THREAD

**Issue by jlaur** [openhab/openhab-addons#11856]
This issue was reported by [Jeewan Bahra](https://community.openhab.org/u/Jeevs) in a [Community post](https://community.openhab.org/t/hd-powerview-binding-connection-error/130552)

All things are offline, but seems to still be controllable.

This is a regression of #11534 which unfortunately was tested only with a v2 hub and with an assumption about API compatibility since Scene Collections are not marked with "V2 only" in the available documentation.

## Expected Behavior
Binding should work correctly for both v1 and v2 hubs.

## Current Behavior
Logged:
`2021-12-24 20:51:41.763 [WARN ] [ternal.handler.HDPowerViewHubHandler] - Error connecting to bridge: java.util.concurrent.ExecutionException: "java.io.EOFException: HttpConnectionOverHTTP@107dadb::SocketChannelEndPoint@108d59a{l=/192.168.1.2:35804,r=/192.168.3.226:80,ISHUT,fill=-,flush=-,to=12/0}{io=0/0,kio=0,kro=1}->HttpConnectionOverHTTP@107dadb(l:/192.168.1.2:35804 <-> r:/192.168.3.226:80,closed=false)=>HttpChannelOverHTTP@1bd77f6(exchange=HttpExchange@183984{req=HttpRequest[GET /api/sceneCollections/ HTTP/1.1]@11cff9e[TERMINATED/null] res=HttpResponse[null 0 null]@a55162[PENDING/null]})[send=HttpSenderOverHTTP@6b9887(req=QUEUED,snd=COMPLETED,failure=null)[HttpGenerator@1d04057{s=START}],recv=HttpReceiverOverHTTP@1acb77c(rsp=IDLE,failure=null)[HttpParser{s=CLOSED,0 of -1}]]"
`

## Possible Solution
 If scene collections are not supported by v1 hub, these channels should not be available and the binding shouldn't try to load them. If it's not possible to determine hub version, the binding should detect inability to load scene collections and stop polling.

## Steps to Reproduce
Simply install and configure binding while having a v1 hub. Problem is in the bridge handler and occurs doing polling, so no actions are needed to reproduce the problem.

**Comment by JeewanBahra** [comment]
@jlaur thanks for creating this bug.

API Responses below:

http:///api/fwversion:
`{"firmware":{"mainProcessor":{"name":"PowerView Hub","revision":1,"subRevision":1,"build":857}}}`

http:///api/sceneCollections/: No response (error)

http:///api/scheduledevents:
`{"scheduledEventIds": [],"scheduledEventData":[]}`

I can see both Scene and Scene Groups in the PowerView app. However, I can only create a Scene - not Scene Group....(not using scene feature)

Automation's are available on V1 but again not actively using this feature as this is being controlled through openHAB

**Comment by jlaur** [comment]
@JeewanBahra - thanks a lot for reporting this, this should make me able to provide a quick fix. For Hub v2 I get the following response (formatted):

```
{
	"firmware": {
		"mainProcessor": {
			"name": "PV Hub2.0",
			"revision": 2,
			"subRevision": 0,
			"build": 1056
		},
		"radio": {
			"revision": 2,
			"subRevision": 0,
			"build": 2610
		}
	}
}
```

So I can use this for correctly identifying hub version, that's good news. For starters I will not be able to provide Scene Group compatibility with Hub v1 as it seems they have renamed/changed that method and I don't own a v1 hub for reverse engineering.

If you are using Android and able and willing to help with technical analysis, perhaps you can install this app:
https://play.google.com/store/apps/details?id=app.greyshirts.sslcapture

It will allow you to start a recording. The only thing you would need to do is open the app freshly (after having flushed it) and go to the scenes activity. Then stop recording and have a look at the requests/responses. That information **may** provide information about how scene groups are fetched through v1 hub. Even though the feature is deprecated now, the app will probably still try to fetch existing scene groups, at least it does for me.

**Comment by jlaur** [comment]
@JeewanBahra - here's a jar file for testing:
https://drive.google.com/file/d/1rYCHiZu_jn0rP8nkS93PQrxG84BuMpsq/view?usp=sharing

I would very much appreciate feedback on:
- Does it work?
- Screenshot of the Hub Thing properties.

For me, properties look like this now with a v2 hub:

![image](https://user-images.githubusercontent.com/19519842/147386576-9eb38282-5b19-49db-acc5-8baee92a855c.png)

**Comment by jlaur** [comment]
@JeewanBahra - secondly I would like to follow up on the scene group issue with v1 hub.
- As you can no longer create any scene groups from the app, do you have any existing ones?
- If you have, would you be able to create a new automation referencing one of those scene groups?
- If possible, after that please send me the response from `http://<hub ip>/api/scheduledevents`

First of all, I would love to be able to support scene groups for v1 hub also. Secondly, automations can reference scene groups. If scene groups are not supported, this can lead to warnings being logged when fetching automations referring to unknown scene groups.

A trace of the app loading scene groups would of course be really awesome, but seeing how automations referencing scene groups look like might also give enough information.

**Comment by JeewanBahra** [comment]
@jlaur - couple of updates;

I installed the Packet Capture app and gone through the process you outlined above. However, I'm not sure I'm understanding this correctly. Is there any way that I can export and attach here?

I hope I've done this correctly, I uninstalled the release version of PowerView and downloaded and placed the jar file into the 'openHAB-addons' folder.

With the above - I see no errors and bridge/rollershutters are all online. 

Properties as follows:

![image](https://user-images.githubusercontent.com/19169425/147387147-5b8d36f1-9bf9-41df-8e44-eb49da52a356.png)

The PowerView app just does not allow me to create a Scene Group. However, can create a normal Scene....

**Comment by jlaur** [comment]
@JeewanBahra - thanks for verifying the fix. From your description you installed it correctly. So if I understand you correctly, this fix resolved your issue?

For the scene group reverse engineering: If you have created a capture from the PowerView:
![image](https://user-images.githubusercontent.com/19519842/147387265-9976935a-a3b6-447a-a522-ae4ce2fc3a99.png)

you should be able to save it to a file:
![image](https://user-images.githubusercontent.com/19519842/147387304-ed43f5fa-430b-4577-b4b9-c61d07346c78.png)

If you have somehow upload/share the contents of that file, that would hopefully reveal how the app fetches scene groups, even if you don't have any. Please make sure that the file doesn't contain any information you wouldn't like to share.

**Comment by JeewanBahra** [comment]
@jlaur - Yes, I can confirm the fix has worked perfectly - Thank you very much for this quick turnaround, it is most appreciated! :)

With regards to scene groups as I've mentioned it doesn't allow me to create - as soon as I click 'New Scene Group' it just closes and goes back to Scenes. 

However, I attach a complete list of steps - one thing I have noticed is that the connection is always closed.

I hope these are helpful in some way shape or form :(

[scenemember.txt](https://github.com/openhab/openhab-addons/files/7775972/scenemember.txt)
[scenecollection.txt](https://github.com/openhab/openhab-addons/files/7775973/scenecollection.txt)
[scene.txt](https://github.com/openhab/openhab-addons/files/7775974/scene.txt)
[scenecollectionmembers.txt](https://github.com/openhab/openhab-addons/files/7775975/scenecollectionmembers.txt)

**Comment by jlaur** [comment]
@JeewanBahra - very useful, thanks!

Can you confirm that:
- `http://<hub ip>/api/sceneCollections/` fails
- ` http://<hub ip>/api/scenecollections/` works

From your trace, this seems to be the difference. And for v2 hub scenecollections will automatically redirect to scene**C**ollections, so this would make it really easy to support both versions.

**Comment by JeewanBahra** [comment]
@jlaur you are absolutely correct! 👯 

`http://<hub ip>/api/sceneCollections/` failed
`http://<hub ip>/api/scenecollections/` works

Output below:

```
{
   "sceneCollectionIds":[
      
   ],
   "sceneCollectionData":[
      
   ]
}
```

**Comment by jlaur** [comment]
@JeewanBahra - I have uploaded a new JAR (same link), can you try this one also? It should support scene groups for v1 hub, but most important is that it should (also) work.

**Comment by JeewanBahra** [comment]
@jlaur downloaded the new JAR and seems to be working as expected - no errors so far :)

**Comment by jlaur** [comment]
> @jlaur downloaded the new JAR and seems to be working as expected - no errors so far :)

@JeewanBahra, thanks a lot for all your assistance and quick feedback. Only because of that was I able to learn enough about the v1 hub to provide a quick fix for this issue. For your information: I created #11857 which directly fixes this incompatibility issue found by your captured packets. This can probably be merged sooner as I doesn't contain all the other changes with firmware information added etc. I'm very happy about this fix because we were able to preserve scene group support and didn't need any firmware version conditions, so thanks again!

**Comment by JeewanBahra** [comment]
@jlaur, thank you for all the help and quick fixes - especially considering today is Christmas day!! Honestly greatly appreciated!

Also, thank you for the clarifications around what fixes have been put in place. :)

Have a great remainder of the festive period :)

**Comment by Arsamin-info** [comment]
> این موضوع توسط [جیوان بهرا](https://community.openhab.org/u/Jeevs) در یک [پست انجمن](https://community.openhab.org/t/hd-powerview-binding-connection-error/130552) گزارش شده است[](https://community.openhab.org/t/hd-powerview-binding-connection-error/130552)
> 
> همه چیز آفلاین است، اما به نظر می رسد هنوز قابل کنترل است.
> 
> این یک رگرسیون از [#11534 است](https://github.com/openhab/openhab-addons/pull/11534) که متأسفانه فقط با یک هاب v2 و با یک فرض در مورد سازگاری API آزمایش شد زیرا مجموعه‌های صحنه با "فقط V2" در اسناد موجود علامت‌گذاری نشده‌اند.
> 
> ## رفتار مورد انتظار
> Binding باید برای هر دو هاب v1 و v2 به درستی کار کند.
> 
> ## رفتار فعلی
> ثبت شده: `2021-12-24 20:51:41.763 [WARN ] [ternal.handler.HDPowerViewHubHandler] - Error connecting to bridge: java.util.concurrent.ExecutionException: "java.io.EOFException: HttpConnectionOverHTTP@107dadb::SocketChannelEndPoint@108d59a{l=/192.168.1.2:35804,r=/192.168.3.226:80,ISHUT,fill=-,flush=-,to=12/0}{io=0/0,kio=0,kro=1}->HttpConnectionOverHTTP@107dadb(l:/192.168.1.2:35804 <-> r:/192.168.3.226:80,closed=false)=>HttpChannelOverHTTP@1bd77f6(exchange=HttpExchange@183984{req=HttpRequest[GET /api/sceneCollections/ HTTP/1.1]@11cff9e[TERMINATED/null] res=HttpResponse[null 0 null]@a55162[PENDING/null]})[send=HttpSenderOverHTTP@6b9887(req=QUEUED,snd=COMPLETED,failure=null)[HttpGenerator@1d04057{s=START}],recv=HttpReceiverOverHTTP@1acb77c(rsp=IDLE,failure=null)[HttpParser{s=CLOSED,0 of -1}]]" `
> 
> ## راه حل ممکن
> If scene collections are not supported by v1 hub, these channels should not be available and the binding shouldn't try to load them. If it's not possible to determine hub version, the binding should detect inability to load scene collections and stop polling.
> 
> ## Steps to Reproduce
> Simply install and configure binding while having a v1 hub. Problem is in the bridge handler and occurs doing polling, so no actions are needed to reproduce the problem.
